drivers/usb_c/tcpc/tcpci.c

```diff
@@ -0,0 +1,370 @@
+		.size = 1,
+	},
+	{
+		.addr = TCPC_REG_TRANSMIT,
+		.name = "TRANSMIT",
+		.size = 1,
+	},
+	{
+		.addr = TCPC_REG_VBUS_VOLTAGE,
+		.name = "VBUS_VOLTAGE",
+		.size = 2,
+	},
+	{
+		.addr = TCPC_REG_VBUS_SINK_DISCONNECT_THRESH,
+		.name = "VBUS_SINK_DISCONNECT_THRESHOLD",
+		.size = 2,
+	},
+	{
+		.addr = TCPC_REG_VBUS_STOP_DISCHARGE_THRESH,
+		.name = "VBUS_STOP_DISCHARGE_THRESHOLD",
+		.size = 2,
+	},
+	{
+		.addr = TCPC_REG_VBUS_VOLTAGE_ALARM_HI_CFG,
+		.name = "VBUS_VOLTAGE_ALARM_HI_CFG",
+		.size = 2,
+	},
+	{
+		.addr = TCPC_REG_VBUS_VOLTAGE_ALARM_LO_CFG,
+		.name = "VBUS_VOLTAGE_ALARM_LO_CFG",
+		.size = 2,
+	},
+	{
+		.addr = TCPC_REG_VBUS_NONDEFAULT_TARGET,
+		.name = "VBUS_NONDEFAULT_TARGET",
+		.size = 2,
+	},
+	{
+		.addr = TCPC_REG_DEV_CAP_3,
+		.name = "DEVICE_CAPABILITIES_3",
+		.size = 2,
+	},
+};
+
+int tcpci_read_reg8(const struct i2c_dt_spec *i2c, uint8_t reg, uint8_t *value)
+{
+	int ret;
+
+	for (int a = 0; a < CONFIG_USBC_TCPC_TCPCI_I2C_RETRIES; a++) {
+		ret = i2c_write_read(i2c->bus, i2c->addr, &reg, sizeof(reg), value, sizeof(*value));
+
+		if (ret == 0) {
+			break;
+		}
+	}
+
+	if (ret != 0) {
+		LOG_ERR(LOG_COMM_ERR_STR, i2c->bus->name, i2c->addr, "r8", reg, *value);
+	}
+
+	return ret;
+}
+
+int tcpci_write_reg8(const struct i2c_dt_spec *i2c, uint8_t reg, uint8_t value)
+{
+	uint8_t buf[2] = {reg, value};
+	int ret;
+
+	for (int a = 0; a < CONFIG_USBC_TCPC_TCPCI_I2C_RETRIES; a++) {
+		ret = i2c_write(i2c->bus, buf, 2, i2c->addr);
+
+		if (ret == 0) {
+			break;
+		}
+	}
+
+	if (ret != 0) {
+		LOG_ERR(LOG_COMM_ERR_STR, i2c->bus->name, i2c->addr, "w8", reg, value);
+	}
+
+	return ret;
+}
+
+int tcpci_update_reg8(const struct i2c_dt_spec *i2c, uint8_t reg, uint8_t mask, uint8_t value)
+{
+	uint8_t old_value;
+	int ret;
+
+	ret = tcpci_read_reg8(i2c, reg, &old_value);
+	if (ret != 0) {
+		return ret;
+	}
+
+	old_value &= ~mask;
+	old_value |= (value & mask);
+
+	ret = tcpci_write_reg8(i2c, reg, old_value);
+
+	return ret;
+}
+
+int tcpci_read_reg16(const struct i2c_dt_spec *i2c, uint8_t reg, uint16_t *value)
+{
+	int ret;
+
+	for (int a = 0; a < CONFIG_USBC_TCPC_TCPCI_I2C_RETRIES; a++) {
+		ret = i2c_write_read(i2c->bus, i2c->addr, &reg, sizeof(reg), value, sizeof(*value));
+
+		if (ret == 0) {
+			*value = sys_le16_to_cpu(*value);
+			break;
+		}
+	}
+
+	if (ret != 0) {
+		LOG_ERR(LOG_COMM_ERR_STR, i2c->bus->name, i2c->addr, "r16", reg, *value);
+	}
+
+	return ret;
+}
+
+int tcpci_write_reg16(const struct i2c_dt_spec *i2c, uint8_t reg, uint16_t value)
+{
+	value = sys_cpu_to_le16(value);
+	uint8_t *value_ptr = (uint8_t *)&value;
+	uint8_t buf[3] = {reg, value_ptr[0], value_ptr[1]};
+	int ret;
+
+	for (int a = 0; a < CONFIG_USBC_TCPC_TCPCI_I2C_RETRIES; a++) {
+		ret = i2c_write(i2c->bus, buf, 3, i2c->addr);
+		if (ret == 0) {
+			break;
+		}
+	}
+
+	if (ret != 0) {
+		LOG_ERR(LOG_COMM_ERR_STR, i2c->bus->name, i2c->addr, "w16", reg, value);
+	}
+
+	return ret;
+}
+
+enum tcpc_alert tcpci_alert_reg_to_enum(uint16_t reg)
```

Comment:
So this function needs to be repeatedly called to get all the pending alerts. Maybe return 0 to signal that no alerts are pending, instead of -1 to signal an error.

Reply:
I wanted these enums to have the same value as bit position in the alert register, because if we have this enum, we can easily clear the bits in the register with BIT(enum_value), where BIT(enum_value - 1) is still valid, but less obvious. But if you think it is better, I can change that.